
<div style="border-radius:10px;
            border : #015a2c solid;
            background-color:#D7D9DB;
           font-size:110%;
           letter-spacing:0.5px;
            text-align: center">
​
<center><h1 style="padding: 25px 0px; color:#015a4c; font-weight: bold; font-family: Cursive">
Amit Project for Data Science Module <br><br>💡💡💡💡💡 <br><br> DEPI_BS_DS_R3  </h1></center>
<center><h3 style="padding-bottom: 25px; color:#025b2c; font-weight: bold; font-style:italic; font-family: Cursive">
Using Project: Trip data Ford go bike  </h3></center>     
​
</div>

# Part I - (Dataset Exploration for Ford GoBike)

### worked by (Manwel Wasfy Georgy)

## Introduction
======================================

.

**This data set information was produced from `Ford GoBike` System** the information in it talked about individual rides that were made in a bike-sharing system for covering San Francisco Bay area.** 

.




> ## And here are the main exploring questions 



   **==================================================**



- What is the structure of your dataset?


- What are the main interesting features in the dataset?


- What features in the dataset supported my investigation for these interesting features?




.

> ### Also here are the investigative questions classified into three types



   **================================================**
   
   


###  <a href='#011'> Univariate Exploration</a> 

- <a href='#1'> 1 ) . What is the distribution of `User type`?</a>  
- <a href='#2'> 2 ) . What is the distribution for the `bike share users on all Trips`?</a>  
- <a href='#3'> 3 ) . What is the distribution for users' `ages`?</a>  
- <a href='#4'> 4 ) . Do the users' ages have `outlier` values?</a>
- <a href='#5'> 5 ) . What is the distribution for the trip duration in `minutes`? </a>
- <a href='#6'> 6) . What is the distribution for `member_gender`? </a>
- <a href='#7'> 7) . What is the distribution for the trip duration in `Hours`?  </a>




##  <a href='#111'> Bivariate Exploration</a> 


- <a href='#8'> 8 ) . What the `average duration` of trip for the categories of `user type`? </a>
- <a href='#9'> 9 ) . How long is the duration of the trip for `each gender` in `minutes`? </a>
- <a href='#10'> 10 ) . How long is the duration of the trip for each `user_type` in `minutes`?  </a>
- <a href='#11'> 11 ) . How the user distribution depending on their `Age` and `user_type` ?</a>
- <a href='#12'>12 ). How the user distribution depending on their `Age` and `member_gender` ?  </a>


##  <a href='#1111'> Multivariate Exploration</a> 


- <a href='#13'> 13 ) . What is the Correlation between `each` variable in the dataset?</a>  
- <a href='#14'> 14 ) . What is the Correlation between `specific numeric variables` in the dataset?</a>  
- <a href='#15'> 15 ) . What is the  Correlation between `data_stat` variables in the dataset?</a>  


##  <a href='#1100'> Conclusions </a> 





# Dataset metadata
Scope: Covers only a single month (February 2019). Do not run cross-month seasonality or annual trend analysis.


# Dataset decomposition
**Temporal** = [duration_sec, start_time, end_time]

**ids&names** = [start_station_id, start_station_name, end_station_id, end_station_name, bike_id] 
    since we are not training a ml model we don't need to drop and could extract facts

**Spatial** = [start_station_latitude, start_station_longitude, end_station_latitude, end_station_longitude]

**User informations** = [user_type, member_birth_year, member_gender]

**Econmic indicator** = [bike_share_for_all_trip]

*nulls exists

    duration_sec => compared to (user_type, birth_year, gender, bike_share_for_all_trip)


    start_time, end_time => already aggregated in duration_sec and doesn't include the hour but could answer smth like 
at which part of the hour do most users start renting, Separate circular trips

    start_station_id, start_station_name, end_station_id, end_station_name => stations with the most rides orignating from, 
stations with the most rides ending at, connected stations

    start_station_latitude, start_station_longitude, end_station_latitude, end_station_longitude => Haversine displacement
distance, Riding speed,

    bike_id => most used bikes, bikes usage compared to (duration, user_type, birth_year, gender)

    user_type, member_gender, bike_share_for_all_trip => category columns with lots of context, should ne used heavily

    bike_share_for_all_trip => Analyze whether low-income program participants use bikes differently

### Dataset Decomposition & Analysis Plan

**Scope:** February 2019 (Cross-month seasonality and annual trend analysis are out of scope).

---

| Category | Columns | Planned Engineering & Analytical Goals |
| :--- | :--- | :--- |
| **Temporal** | `duration_sec`<br>`start_time`<br>`end_time` | • Convert duration to minutes; apply log scaling to handle extreme skews.<br>• Bin start minutes (`0–10`, `10–20`, etc.) to evaluate transit connection patterns within the hour.<br>• Cross-tabulate duration against demographics, user type, and equity program enrollment. |
| **Network & Stations** | `start_station_id`<br>`start_station_name`<br>`end_station_id`<br>`end_station_name` | • Identify top origin, destination, and paired travel corridors.<br>• Calculate net station flow ($\text{Arrivals} - \text{Departures}$) to pinpoint bike surplus and deficit hubs.<br>• Flag circular round-trips (`start_station_id == end_station_id`) vs. one-way transit. |
| **Spatial** | `start_station_latitude`<br>`start_station_longitude`<br>`end_station_latitude`<br>`end_station_longitude` | • Calculate straight-line displacement via the Haversine formula.<br>• Approximate speed ($\text{Distance} / \text{Duration}$) to detect non-movement outliers or leisure cruising.<br>• Map geographic trip clusters across San Francisco, Oakland, and San Jose. |
| **Demographics** | `user_type`<br>`member_birth_year`<br>`member_gender` | • Derive user age (`2019 - member_birth_year`) and group into demographic cohorts (e.g., `<25`, `25–34`, `35–54`, `55+`).<br>• Filter implausible birth years (e.g., ages $>80$).<br>• Compare ride duration across gender identities and user subscription models (`Subscriber` vs. `Customer`). |
| **Equity Program** | `bike_share_for_all_trip` | • Measure equity program adoption among eligible riders.<br>• Evaluate whether low-income program participants differ in trip lengths, repeat usage, or primary station hubs. |
| **Fleet Tracking** | `bike_id` | • Rank bikes by cumulative usage (ride counts and total operating hours).<br>• Identify heavily utilized inventory for proactive maintenance flags. |

---

### Data Hygiene & Null Strategy

* **Station Metadata (`start_station_id`, `end_station_id`):** 197 missing records (~0.1%). Exclude from station-level routing analyses or tag as dockless/off-grid trips; keep for aggregate duration metrics.
* **Demographics (`member_birth_year`, `member_gender`):** 8,265 missing records (~4.5%), concentrated in casual `Customer` trips. Segment demographic analyses exclusively to authenticated `Subscriber` data rather than imputing artificial values.

# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Import Libraries</p>

In [ ]:
import numpy as np
import pandas as pd
import os
from pathlib import Path
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:
# Load dataset (CSV lives next to this notebook in Project/Solo)
csv_path = Path("fordgobike-tripdataFor201902.csv")
if not csv_path.exists():
    csv_path = Path("/home/manwelw/depi/Depi/data-analysis/Project/Solo/fordgobike-tripdataFor201902.csv")
df = pd.read_csv(csv_path)
df.head()


In [59]:
# here we make overview in high-level for data shape and composition
df.shape

(183416, 16)

In [60]:
## check each var  datatype 
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 183416 entries, 0 to 183415
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   duration_sec             183416 non-null  int64  
 1   start_time               183416 non-null  str    
 2   end_time                 183416 non-null  str    
 3   start_station_id         183219 non-null  float64
 4   start_station_name       183219 non-null  str    
 5   start_station_latitude   183416 non-null  float64
 6   start_station_longitude  183416 non-null  float64
 7   end_station_id           183219 non-null  float64
 8   end_station_name         183219 non-null  str    
 9   end_station_latitude     183416 non-null  float64
 10  end_station_longitude    183416 non-null  float64
 11  bike_id                  183416 non-null  int64  
 12  user_type                183416 non-null  str    
 13  member_birth_year        175151 non-null  float64
 14  member_gender  

In [61]:
# descriptive statistics for numeric variables
df.describe().T

,count,mean,std,min,25%,50%,75%,max
duration_sec,183416.0,726.067311,1794.371831,61.000000,325.000000,514.000000,796.000000,85444.000000
start_station_id,183219.0,138.590365,111.778938,3.000000,47.000000,104.000000,239.000000,398.000000
start_station_latitude,183416.0,37.771221,0.099586,37.317298,37.770083,37.780760,37.797280,37.880222
start_station_longitude,183416.0,-122.352661,0.117100,-122.453705,-122.412408,-122.398285,-122.286533,-121.874119
end_station_id,183219.0,136.249401,111.514916,3.000000,44.000000,100.000000,235.000000,398.000000
end_station_latitude,183416.0,37.771425,0.099494,37.317298,37.770407,37.781010,37.797320,37.880222
end_station_longitude,183416.0,-122.352247,0.116677,-122.453705,-122.411726,-122.398279,-122.288045,-121.874119
bike_id,183416.0,4472.890260,1664.386068,11.000000,3777.000000,4958.000000,5502.000000,6645.000000
member_birth_year,175151.0,1984.806504,10.116589,1878.000000,1980.000000,1987.000000,1992.000000,2001.000000


In [62]:
df.dtypes

duration_sec                 int64
start_time                     str
end_time                       str
start_station_id           float64
start_station_name             str
start_station_latitude     float64
start_station_longitude    float64
end_station_id             float64
end_station_name               str
end_station_latitude       float64
end_station_longitude      float64
bike_id                      int64
user_type                      str
member_birth_year          float64
member_gender                  str
bike_share_for_all_trip        str
dtype: object



> ### What is the structure of your dataset?





- The structure of the dataset contains 16 columns and 183416 rows.
 
 
- Mixed types: 9 numeric ( durations, ids, coordinates, bike_id, birth_year ) and 7 string/categorical ( times, station names, user_type, gender, equity flag ) , and here are **dtypes summary :**

        duration_sec                 int64
        start_time                     str
        end_time                       str
        start_station_id           float64
        start_station_name             str
        start_station_latitude     float64
        start_station_longitude    float64
        end_station_id             float64
        end_station_name               str
        end_station_latitude       float64
        end_station_longitude      float64
        bike_id                      int64
        user_type                      str
        member_birth_year          float64
        member_gender                  str
        bike_share_for_all_trip        str
        dtype: object

> ### What are the main interesting features in the dataset?




 **We got three main feature areas which are :**
 
 
 
 
- Temporal
 
 
 
- Demographics
 
  
 
  
- Network & Stations




> ### What features in the dataset supported my investigation for these interesting features?



- start_station_id, start_station_name

- end_station_id, end_station_name

- bike_id

- user_type (Customer, Subscriber)

- member_gender (Male, Female, Other)

- bike_share_for_all_trip (Yes, No)

## <mark>The above result show some missing values exist in some columns So let's display them </mark>

In [ ]:
df.isna().sum()
# NOTE: member_birth_year/member_gender nulls are NOT only Customers.
# Cell 21 check shows ~3.2k Customers AND ~5k Subscribers missing demographics,
# so do not attribute all missing demographics to casual rides.
# Station nulls (197) will be dropped; demographic nulls will be retained (see below).




> ## Missing-data strategy

- **Station columns** (`start_station_id/name`, `end_station_id/name`, 197 rows, ~0.1%): drop via `dropna(subset=...)`. Too few to impute; needed for routing analysis.

- **Demographics** (`member_birth_year`, `member_gender`, 8,265 rows, ~4.5%): **do NOT impute** with mean/mode — that would fabricate ages/genders. Keep as NaN and segment demographic analyses to non-null rows.

- **Duplicates** (4 rows): inspect, then drop (see duplicate cell below).


In [ ]:
# Drop the 197 rows missing station id/name (needed for origin/destination analysis).
# This is a drop, not a fill: imputing station names would fabricate routes.
cols_to_drop = ["start_station_id", "start_station_name", "end_station_id", "end_station_name"]
df = df.dropna(subset=cols_to_drop)


In [65]:
#check the Nulls 
df.isna().sum()

duration_sec                  0
start_time                    0
end_time                      0
start_station_id              0
start_station_name            0
start_station_latitude        0
start_station_longitude       0
end_station_id                0
end_station_name              0
end_station_latitude          0
end_station_longitude         0
bike_id                       0
user_type                     0
member_birth_year          8263
member_gender              8263
bike_share_for_all_trip       0
dtype: int64

### **And with a proper category for other (member gender)**

In [66]:
df['member_gender'].value_counts()

member_gender
Male      130504
Female     40805
Other       3647
Name: count, dtype: int64

In [67]:
df['user_type'].value_counts()

user_type
Subscriber    163418
Customer       19801
Name: count, dtype: int64

In [68]:
##  still got a missing value in the member_gender, so let's explore this column more 
df[ (df['user_type'] == 'Customer') & (df['member_gender'].isna())].count()

duration_sec               3235
start_time                 3235
end_time                   3235
start_station_id           3235
start_station_name         3235
start_station_latitude     3235
start_station_longitude    3235
end_station_id             3235
end_station_name           3235
end_station_latitude       3235
end_station_longitude      3235
bike_id                    3235
user_type                  3235
member_birth_year             0
member_gender                 0
bike_share_for_all_trip    3235
dtype: int64

In [ ]:
# Demographic nulls: quantify by user_type instead of filling with fake values.
print("Remaining birth_year nulls:", int(df["member_birth_year"].isna().sum()))
print("Remaining gender nulls:", int(df["member_gender"].isna().sum()))
print(df.groupby("user_type", observed=True)[["member_birth_year", "member_gender"]].apply(lambda g: g.isna().sum()))
# Strategy: keep NaN; filter to non-null only inside age/gender plots (e.g. df.dropna(subset=['member_age']))


In [ ]:
# Sanity check: Customers missing gender vs Subscribers missing gender
print("Customers missing gender:", int(((df['user_type'] == 'Customer') & (df['member_gender'].isna())).sum()))
print("Subscribers missing gender:", int(((df['user_type'] == 'Subscriber') & (df['member_gender'].isna())).sum()))


In [ ]:
## Result: ~3.2k Customers + ~5k Subscribers lack demographics, so missingness is not 'just casual rides'.
print("Total rows after station drop:", len(df))


> ## Station nulls are resolved (197 rows dropped). Demographic nulls (~8.3k) are intentionally retained as NaN — see strategy above.

### Now let's check duplicate value

In [ ]:
## Check duplicates, inspect, then drop exact duplicate rows
print("Duplicate rows:", int(df.duplicated().sum()))
# display(df[df.duplicated(keep=False)])  # uncomment to inspect the 2 pairs
df = df.drop_duplicates().reset_index(drop=True)
print("Rows after dedup:", len(df))


> ## Duplicates (4 rows / 2 pairs) dropped via `drop_duplicates()` and index reset.


### Now let's check the category columns to make more cleaning dataset 

In [73]:
### check the category columns that included objects
cat_cols = ['user_type', 'member_gender', 'bike_share_for_all_trip']

for col in cat_cols:
    print(f"--- Column: {col} ---")
    print(f"Unique values: {df[col].unique()}")
    print(df[col].value_counts(dropna=False))
    print()

--- Column: user_type ---
Unique values: <StringArray>
['Customer', 'Subscriber']
Length: 2, dtype: str
user_type
Subscriber    163418
Customer       19801
Name: count, dtype: int64

--- Column: member_gender ---
Unique values: <StringArray>
['Male', nan, 'Other', 'Female']
Length: 4, dtype: str
member_gender
Male      130504
Female     40805
NaN         8263
Other       3647
Name: count, dtype: int64

--- Column: bike_share_for_all_trip ---
Unique values: <StringArray>
['No', 'Yes']
Length: 2, dtype: str
bike_share_for_all_trip
No     165872
Yes     17347
Name: count, dtype: int64



 > ## From all previous results above I found  :
 
 
 
 **=========================================================**
 
 

> 1 ) **unique Value** as in ((`member_gender`)) : Not significant , so won't clean it



> 2 ) **Incorrect data type** as in (( `user_type` &&	`member_gender`)) :  significant change , so I need to clean it.


> 3 ) **Good factor** for marketing work as in ((`member_birth_year`)) : so I need to clean it.

In [74]:
##  step  2 ) Incorrect data type as in (( `user_type` &&	`member_gender`)) :  significant change , so  need to clean it.
df['user_type'] = df['user_type'].astype('category')
df['member_gender'] = df['member_gender'].astype('category')
df['bike_share_for_all_trip'] = df['bike_share_for_all_trip'].astype('category')

In [ ]:
## check your work
print(df[['user_type','member_gender','bike_share_for_all_trip']].dtypes)


In [ ]:
## Step 3) Marketing feature: derive rider age from birth year
## Bounds 16-95: 16 = minimum plausible rider age; 95 = generous upper bound.
## Ages >80 flagged as suspicious per analysis plan; >95 treated as data-entry error (e.g. 1878 -> age 141).
df['member_age'] = 2019 - df['member_birth_year']
print(df['member_age'].describe())
# max of 141 is an outlier for sure


In [ ]:
## Cap implausible ages to NaN (adds ~75 outliers to the ~8263 demographic NaNs)
df['member_age'] = df['member_age'].where((df['member_age'] >= 16) & (df['member_age'] <= 95), np.nan)
print("member_age NaNs (original nulls + ~75 age outliers):", int(df['member_age'].isna().sum()))


In [ ]:
## Convert age to nullable Int64 (plain int cannot hold NaN)
df['member_age'] = df['member_age'].astype('Int64')
print(df['member_age'].dtype)
print(df['member_age'].describe())


In [ ]:
## Helper: average / median rider age (skips NaN)
def mean_median_age(s):
    return float(s.mean(skipna=True)), float(s.median(skipna=True))
mean_age, median_age = mean_median_age(df['member_age'].astype('float'))
print(f"Mean age: {mean_age:.1f}, Median age: {median_age:.1f} (non-null n={int(df['member_age'].notna().sum())})")


In [ ]:
## Trip duration in minutes (more usable than seconds for marketing)
df['duration_min'] = df['duration_sec'] / 60.0
print(df[['duration_sec', 'duration_min']].describe().T)


In [ ]:
## Trip duration in hours (for long-trip / outlier review)
df['duration_hr'] = df['duration_sec'] / 3600.0
print(df[['duration_sec', 'duration_min', 'duration_hr']].describe().T)


In [ ]:
## Verify new columns: dtypes, null counts, sanity ranges
print(df[['member_age', 'duration_min', 'duration_hr']].dtypes)
print(df[['member_age', 'duration_min', 'duration_hr']].isna().sum())
print("duration_min > 30 min share:", float((df['duration_min'] > 30).mean()))



# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Univariate Exploration</p>
# <a id="011"> Univariate Exploration</a>



> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.




# <a id="1"> 1 ) . What is the distribution for User type?</a>

In [ ]:
## Q1: distribution of user_type
vc = df['user_type'].value_counts()
print(vc)
print((vc / vc.sum() * 100).round(1).astype(str) + ' %')
plt.figure(figsize=(6,4))
sns.countplot(data=df, x='user_type', order=['Subscriber','Customer'])
plt.title('Ride counts by user_type (Feb 2019)')
plt.xlabel('User type'); plt.ylabel('Rides')
plt.tight_layout(); plt.show()


> ### <mark>Your analysis and exploration here </mark>

# <a id="2"> 2 ) What is the distribution for the bike share users on all Trips?</a>

> ### <mark>Your analysis and exploration here </mark>

In [ ]:
## Q2: distribution of bike_share_for_all_trip (equity program)
vc = df['bike_share_for_all_trip'].value_counts()
print(vc)
print((vc / vc.sum() * 100).round(2).astype(str) + ' %')
print(pd.crosstab(df['user_type'], df['bike_share_for_all_trip'], normalize='index').round(3).to_string())
plt.figure(figsize=(6,4))
sns.countplot(data=df, x='bike_share_for_all_trip', order=['No','Yes'])
plt.title('Rides by Bike-Share-for-All enrollment')
plt.xlabel('Equity trip'); plt.ylabel('Rides')
plt.tight_layout(); plt.show()


# <a id="3"> 3 ) What is the distribution for users' ages? </a>

> ### <mark>Your analysis and exploration here </mark>

In [ ]:
## Q3: distribution of rider ages (right-skewed; mean > median)
a = df['member_age'].astype('float')
print(a.describe().to_string())
print('Mean / median:', round(float(a.mean()),1), '/', round(float(a.median()),1))
plt.figure(figsize=(8,4))
sns.histplot(a.dropna(), bins=50, kde=True)
plt.axvline(a.median(), color='red', linestyle='--', label='median')
plt.title('Age distribution (non-null riders)')
plt.xlabel('Age'); plt.ylabel('Rides'); plt.legend()
plt.tight_layout(); plt.show()



# <a id="4">  4 ) . Do the users' ages have outlier values?</a>


> ### <mark>Your analysis and exploration here </mark>

In [ ]:
## Q4: do ages have outliers? boxplot + IQR rule + cap audit
a = df['member_age'].astype('float').dropna()
q1, q3 = a.quantile(0.25), a.quantile(0.75)
iqr = q3 - q1
lo, hi = q1 - 1.5*iqr, q3 + 1.5*iqr
print(f'IQR bounds: [{lo:.1f}, {hi:.1f}]')
print('Beyond IQR:', int(((a < lo) | (a > hi)).sum()), '| Age > 80:', int((a > 80).sum()))
plt.figure(figsize=(7,3))
sns.boxplot(x=a)
plt.title('Age outliers (boxplot, non-null)')
plt.xlabel('Age')
plt.tight_layout(); plt.show()
# NOTE: raw max was 141 (birth year 1878); values >95 already capped to NaN in cleaning.


# <a id="5">  5 ) . What is the distribution for the trip duration in minutes? </a>

In [ ]:
## Q5 stats: trip duration in minutes
print(df['duration_min'].describe().to_string())
print('Share > 30 min:', round(float((df['duration_min'] > 30).mean()*100), 2), '%')
print('95th / 99th pct:', df['duration_min'].quantile([0.95, 0.99]).to_dict())


In [ ]:
## Q5 viz: duration_min is heavily right-skewed -> clip at 99th pct for readability
p99 = float(df['duration_min'].quantile(0.99))
plt.figure(figsize=(8,4))
sns.histplot(df['duration_min'].clip(upper=p99), bins=60)
plt.axvline(df['duration_min'].median(), color='red', linestyle='--', label='median')
plt.title(f'Trip duration (min), clipped at 99th pct={p99:.1f}')
plt.xlabel('Minutes'); plt.ylabel('Rides'); plt.legend()
plt.tight_layout(); plt.show()


> ### <mark>Your analysis and exploration here </mark>


# <a id="6"> 6 ) . What is the distribution for member_gender?</a>

In [ ]:
## Q6: distribution of member_gender (plot non-null; report NaN separately)
print(df['member_gender'].value_counts(dropna=False))
print('Missing gender:', int(df['member_gender'].isna().sum()))
plt.figure(figsize=(6,4))
sns.countplot(data=df.dropna(subset=['member_gender']), x='member_gender', order=['Male','Female','Other'])
plt.title('Rides by member_gender (non-null only)')
plt.xlabel('Gender'); plt.ylabel('Rides')
plt.tight_layout(); plt.show()


> ### <mark>Your analysis and exploration here </mark>

# <a id="7">    7 )  . What is the distribution for the trip duration in Hours?  </a>

In [ ]:
## Q7: duration in hours (same shape as minutes / 60; long-tail outliers visible)
print(df['duration_hr'].describe().to_string())
p99h = float(df['duration_hr'].quantile(0.99))
plt.figure(figsize=(8,4))
sns.histplot(df['duration_hr'].clip(upper=p99h), bins=60)
plt.title(f'Trip duration (hours), clipped at 99th pct={p99h:.2f}h')
plt.xlabel('Hours'); plt.ylabel('Rides')
plt.tight_layout(); plt.show()


> ### <mark>Your analysis and exploration here </mark>

### The variables distribution:


> The distribution of duration present that most of the trips were less than 30 minutes ( we can notice that from the minutes and the hour charts )


> And the Age distribution is ( right-skewed / positively skewed ) as most users age are  22 - 39


> Most users are subscribers, not Customers and  Males as shown in the charts above.



### Were there any interesting and outliers 


- We got Age outliers for for more than 100 .



- The distribution of duration was strange where the minutes and the hour charts were 30 minutes or less.

# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Bivariate Exploration</p>
# <a id="111"> Bivariate Exploration</a>

Now let's investigate relationships between pairs of variables in our data.

# <a id="8"> 8) . What the average duration of trip for the categories of user type?</a>

In [ ]:
## Q8: average trip duration by user_type
t = df.groupby('user_type', observed=True)['duration_min'].agg(['mean','median','count'])
print(t.round(2).to_string())
plt.figure(figsize=(6,4))
sns.barplot(data=df, x='user_type', y='duration_min', estimator='mean', errorbar=None, order=['Subscriber','Customer'])
plt.title('Mean trip duration (min) by user_type')
plt.xlabel('User type'); plt.ylabel('Mean minutes')
plt.tight_layout(); plt.show()


# Make visualization to easy get insight


> ### <mark>Your analysis and exploration here </mark>

# <a id="9">    9 ) . How long is the duration of the trip for each gender in minutes?  </a>

In [ ]:
## Q9: trip duration (min) for each gender — medians + capped boxplot
print(df.groupby('member_gender', observed=True)['duration_min'].agg(['mean','median','count']).round(2).to_string())
cap = float(df['duration_min'].quantile(0.99))
plt.figure(figsize=(7,4))
sns.boxplot(data=df.dropna(subset=['member_gender']), x='member_gender', y='duration_min', order=['Male','Female','Other'])
plt.ylim(0, cap)
plt.title(f'Duration (min) by gender (y capped at 99th pct={cap:.1f})')
plt.xlabel('Gender'); plt.ylabel('Minutes')
plt.tight_layout(); plt.show()


# Make visualization to easy get insight


> ### <mark>Your analysis and exploration here </mark>

# <a id="10 "> 10 ) . How long is the duration of the trip for each user_type in minutes?  </a>

In [ ]:
## Q10: trip-duration distribution per user_type (shape, not just the mean)
print(df.groupby('user_type', observed=True)['duration_min'].agg(['mean','median','count']).round(2).to_string())
cap = float(df['duration_min'].quantile(0.99))
plt.figure(figsize=(7,4))
sns.boxplot(data=df, x='user_type', y='duration_min', order=['Subscriber','Customer'])
plt.ylim(0, cap)
plt.title(f'Duration (min) by user_type (y capped at 99th pct={cap:.1f})')
plt.xlabel('User type'); plt.ylabel('Minutes')
plt.tight_layout(); plt.show()


> ### <mark>Your analysis and exploration here </mark>

# <a id="11">  11 ) . How the user distribution depending on their Age and user_type ? </a>

In [ ]:
## Q11: age distribution split by user_type + cohort crosstab
age = df.dropna(subset=['member_age']).copy()
age['age_cohort'] = pd.cut(age['member_age'].astype(int), bins=[0,25,35,55,200], labels=['<25','25-34','35-54','55+'])
print(pd.crosstab(age['age_cohort'], age['user_type'], normalize='columns').round(3).to_string())
plt.figure(figsize=(8,4))
sns.histplot(data=age, x='member_age', hue='user_type', bins=40, element='step', stat='density', common_norm=False)
plt.title('Age distribution by user_type (density)')
plt.xlabel('Age'); plt.ylabel('Density')
plt.tight_layout(); plt.show()


> ### <mark>Your analysis and exploration here </mark>

#  <a id="12"> 12 ). How the user distribution depending on their Age and member_gender ?</a>

In [ ]:
## Q12: age distribution split by member_gender
ageg = df.dropna(subset=['member_age','member_gender']).copy()
print(ageg.groupby('member_gender', observed=True)['member_age'].agg(['mean','median','count']).round(1).to_string())
plt.figure(figsize=(8,4))
sns.histplot(data=ageg, x='member_age', hue='member_gender', bins=40, element='step', stat='density', common_norm=False)
plt.title('Age distribution by member_gender (density)')
plt.xlabel('Age'); plt.ylabel('Density')
plt.tight_layout(); plt.show()


> ### <mark>Your analysis and exploration here </mark>

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> ...................


### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> ....................

# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Multivariate Exploration</p>
# <a id="1111"> Multivariate Exploration </a>

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

In [93]:
## check your work 



# <a id="13">  13 ) . What is the  Correlation between each variable in the dataset?</a>

In [ ]:
## Q13: correlation across all numeric columns (timestamps excluded: corrupt in source)
num = df.select_dtypes(include='number').copy()
num['member_age'] = num['member_age'].astype('float')
corr = num.corr(numeric_only=True)
print(corr.round(2).to_string())
plt.figure(figsize=(10,8))
sns.heatmap(corr, annot=False, cmap='coolwarm', center=0, square=True)
plt.title('Correlation: all numeric columns')
plt.tight_layout(); plt.show()




> ### <mark>Your analysis and exploration here </mark>

Example usage
- There is `no Correlation` for any variable with time ( `duration sec /minutes/hours`)



# <a id="14"> 14 )  .  What is the  Correlation between `specific numeric variables` in the dataset?</a>


In [ ]:
## Q14: focused numeric set (avoids 16x16 clutter)
cols = ['duration_sec','member_age','member_birth_year','bike_id','start_station_latitude','start_station_longitude']
sub = df[cols].copy()
sub['member_age'] = sub['member_age'].astype('float')
corr = sub.corr(numeric_only=True)
print(corr.round(3).to_string())
plt.figure(figsize=(7,5))
sns.heatmap(corr, annot=True, fmt='.2f', cmap='coolwarm', center=0, square=True)
plt.title('Correlation: selected numeric variables')
plt.tight_layout(); plt.show()
# NOTE: birth_year vs member_age = -1.0 by construction (age = 2019 - birth_year).


> ### <mark>Your analysis and exploration here </mark>

Example usage
- There is `no Correlation` for any variable with time ( `duration sec /minutes`)




- 




- 




# <a id="15">  15 ) What is the  Correlation between data_stat variables in the dataset?</a>


In [ ]:
"""Q15: duration/age descriptives are perfectly linked by construction (min=sec/60, hr=sec/3600)."""
cols = ['duration_sec','duration_min','duration_hr','member_age','member_birth_year']
sub = df[cols].copy()
sub['member_age'] = sub['member_age'].astype('float')
corr = sub.corr(numeric_only=True)
print((corr*100).round(1).astype(str) + ' %')
plt.figure(figsize=(6,5))
sns.heatmap(corr, annot=True, fmt='.2f', cmap='coolwarm', center=0, square=True)
plt.title('Correlation: duration + age block (% in console)')
plt.tight_layout(); plt.show()


> ### <mark>The above charts show how Correlation between specific numeric variables in the dataset using heatmap plots  </mark>



**Actually the `heat map`  display the above ideas in clear percentages**



- There is `no Correlation` for any variable with time ( `duration sec /minutes`)




- The `birth_year vs member_age r=-1` is definitional (`age = 2019 - birth_year`), not a finding — exclude one of them from correlation matrices.




- There is `no significant Correlation` for any variable with `bike_id `



# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Conclusions</p>
# <a id="1100"> Conclusions </a>

## Key findings (Feb 2019, n=183,215 after cleaning)

- **Subscriber system:** 89.2% Subscriber (163,414) vs 10.8% Customer (19,801). Core 25-34 cohort is ~47-49% of rides in both groups; 55+ is small (Subscribers 4.6%, Customers 3.4%).

- **Gender split (non-null):** Male 130,500 / Female 40,805 / Other 3,647, plus 8,263 missing. Female riders are youngest on average (mean 33.2, median 31) vs Male (34.4 / 32) and Other (35.9 / 34).

- **Age:** mean 34.2, median 32.0, right-skewed. Raw max 141 (birth year 1878) is a data-entry error; 75 values >95 were capped to NaN, leaving 117 riders >80. IQR bounds [9.0, 57.0] flag 5,706 statistical outliers, but most are legitimate older riders — only >95 was treated as error.

- **Trip length:** mean 12.1 min, median 8.6 min; 95th pct 26.2 min, 99th pct 57.6 min; only 3.68% of trips exceed 30 min. Max 1,424 min (23.7 h) indicates overnight/leisure/non-return outliers, not typical use.

- **Who rides longer:** Customers average 23.9 min (median 13.2) vs Subscribers 10.7 min (median 8.2) — casual users take ~2x longer trips. By gender, Other has the highest mean (16.6 min), then Female (13.0), then Male (11.2); medians are closer (9.5 / 9.3 / 8.2).

- **Equity program:** 9.5% of rides enrolled (17,346). Enrollment is 0% among Customers and 10.6% among Subscribers, so the program reaches committed members, not casual riders.

- **Correlations:** effectively no linear link between duration and demographics, bike_id, or station coordinates (|r| < 0.05). Start/end coordinates correlate ~0.99 (short hops); longitude-latitude ~-0.67 reflects Bay Area geography. `birth_year vs age = -1.0` is definitional (`age = 2019 - birth_year`), and `sec/min/hr` are 100% linked by construction — neither is a finding.

## Limitations

- **Single month (Feb 2019):** no seasonality, annual-trend, or weather/holiday analysis possible.

- **Corrupt timestamps in source CSV:** `start_time`/`end_time` contain only `MM:SS.ms` (e.g. `32:10.1`), no date/hour, so time-of-day, rush-hour, and dwell-time analysis were out of scope and left untouched.

- **Missing demographics (~4.5%, 8,263 rows + 75 age caps = 8,338 age NaNs):** present in both Customers (3,235) and Subscribers (5,028), so results by age/gender describe only reporting riders; values were retained as NaN, never imputed.

- **Duration outliers retained:** extreme long trips inflate means — medians and 99th-pct clipping were used for plots, but no trips were deleted for being long.

- **Station coordinates are dock locations, not GPS traces:** true route distance, speed, and circular-trip detection need Haversine displacement on top of duration, not done here.

- **Equity flag only:** `bike_share_for_all_trip` records enrollment, not income or eligibility, so behavioral differences are descriptive, not causal.
